Disabling users in WebUI

MorrisJobke opened the pull request:
- #413 rebased on master - by @Faldon 
- I reviewed the layout and improved it a bit
- I will continue reviewing it and read through all comments in #413 and try to fix them

Commit:
Implemented visual feedback if a user is disabled in admin user menu.

Implemented visuals for enabling/disabling user from admin user list.
Added the controller functions for enabling/disabling a user.

Added the route for changing user status (enabled/disabled) and added an additional route handler in the user controller.
 Finished the visuals to reflect current user status and changed user status respectively.

Changed the single icon for enabling/disabling a user into a menu where deletion and state toggling of a user is selectable.

Added displaying of disabled user count.
Improved style of user action menu.

Added proper counting of disabled users.
Removed visual indicator for disabled users.

Moved pseudo-group detection for disabled users from frontend to the controller.
Changed units for newly introduced css values from em to px.
Removed unnecessary png and optimized svg with scour.
Changed the userlist template to display the user action menu with correct width.

Style fixes for better readability and coding style conformity.

Changed the icons for enabling, disabling and deleting a user in the action menu.

MorrisJobke committed:
improve layout

MorrisJobke commented:
It now looks like this:

<img width="294" alt="bildschirmfoto 2016-09-01 um 16 53 20" src="https://cloud.githubusercontent.com/assets/245432/18172285/1bdb3e1c-7065-11e6-9984-e4ad741db34c.png">

MorrisJobke commented:
I tested this and it works quite well :+1: from me

@rullzer @nickvergessen @LukasReschke @mar1u5 @nextcloud/designers Please review

MorrisJobke commented:
```
There were 13 failures:

1) Tests\Settings\Controller\UsersControllerTest::testIndexAdmin
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\DataResponse Object (
     'data' => Array (
-        0 => Array (...)
-        1 => Array (...)
-        2 => Array (...)
     )
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 200
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:266

2) Tests\Settings\Controller\UsersControllerTest::testIndexSubAdmin
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\DataResponse Object (
     'data' => Array (
-        0 => Array (...)
-        1 => Array (...)
-        2 => Array (...)
     )
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 200
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:473

3) Tests\Settings\Controller\UsersControllerTest::testIndexWithSearch
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\DataResponse Object (
     'data' => Array (
-        0 => Array (...)
-        1 => Array (...)
-        2 => Array (...)
     )
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 200
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:635

4) Tests\Settings\Controller\UsersControllerTest::testIndexWithBackend
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\DataResponse Object (
     'data' => Array (
-        0 => Array (...)
     )
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 200
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:711

5) Tests\Settings\Controller\UsersControllerTest::testCreateSuccessfulWithoutGroupAdmin
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
         'backend' => 'bar'
-        'lastLogin' => null
+        'lastLogin' => 0
         'displayname' => null
         'quota' => null
         'subadmin' => Array ()
-        'email' => null
+        'email' => ''
         'isRestoreDisabled' => false
         'isAvatarAvailable' => true
+        'isEnabled' => null
     )
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 201
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:783

6) Tests\Settings\Controller\UsersControllerTest::testCreateSuccessfulWithoutGroupSubAdmin
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
         'subadmin' => Array ()
-        'email' => null
+        'email' => ''
         'isRestoreDisabled' => false
         'isAvatarAvailable' => true
+        'isEnabled' => null
     )
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 201
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:872

7) Tests\Settings\Controller\UsersControllerTest::testCreateSuccessfulWithGroupAdmin
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
         'backend' => 'bar'
-        'lastLogin' => null
+        'lastLogin' => 0
         'displayname' => null
         'quota' => null
         'subadmin' => Array ()
-        'email' => null
+        'email' => ''
         'isRestoreDisabled' => false
         'isAvatarAvailable' => true
+        'isEnabled' => null
     )
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 201
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:955

8) Tests\Settings\Controller\UsersControllerTest::testCreateSuccessfulWithGroupSubAdmin
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
         'subadmin' => Array ()
-        'email' => null
+        'email' => ''
         'isRestoreDisabled' => false
         'isAvatarAvailable' => true
+        'isEnabled' => null
     )
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 201
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:1051

9) Tests\Settings\Controller\UsersControllerTest::testRestorePossibleWithoutEncryption
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
     'backend' => 'OC_User_Database'
-    'email' => null
+    'email' => ''
     'isRestoreDisabled' => false
     'isAvatarAvailable' => true
+    'isEnabled' => null
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:1567

10) Tests\Settings\Controller\UsersControllerTest::testRestorePossibleWithAdminAndUserRestore
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
     'backend' => 'OC_User_Database'
-    'email' => null
+    'email' => ''
     'isRestoreDisabled' => false
     'isAvatarAvailable' => true
+    'isEnabled' => null
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:1616

11) Tests\Settings\Controller\UsersControllerTest::testRestoreNotPossibleWithoutAdminRestore
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
     'backend' => 'OC_User_Database'
-    'email' => null
+    'email' => ''
     'isRestoreDisabled' => true
     'isAvatarAvailable' => true
+    'isEnabled' => null
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:1646

12) Tests\Settings\Controller\UsersControllerTest::testRestoreNotPossibleWithoutUserRestore
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
     'backend' => 'OC_User_Database'
-    'email' => null
+    'email' => ''
     'isRestoreDisabled' => true
     'isAvatarAvailable' => true
+    'isEnabled' => null
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:1697

13) Tests\Settings\Controller\UsersControllerTest::testNoAvatar
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
     'backend' => 'OC_User_Database'
-    'email' => null
+    'email' => ''
     'isRestoreDisabled' => false
     'isAvatarAvailable' => false
+    'isEnabled' => null
 )

/drone/src/github.com/nextcloud/server/tests/Settings/Controller/UsersControllerTest.php:1723
```

Faldon commented:
@MorrisJobke Sorry, was out of town for a few days. I'm glad someone more drone experienced does the fixes. ;-)

rullzer commented:
- [ ] SVGs need to be minimized
- [ ] SVGs need proper viewBox

rullzer commented on settings/Controller/UsersController.php:
```diff
@@ -245,15 +241,18 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 		if ($this->isAdmin) {
-
-			if($gid !== '') {
+			if($gid !== '' && $gid !== 'disabledUsers') {
```
This will explode if anybody ever creates a group `disabledUsers`

MorrisJobke replied:
Then name it NextcloudDisabledUsers

rullzer replied:
Why not just another parameter $disabledUsers that defaults to false?

MorrisJobke replied:
Ah yes ... makes sense then 🙈 It's early in the morning 😝

nickvergessen replied:
Well we have the same problem with the group "everyone"

rullzer committed:
Fix unit tests

rullzer commented:
* Rebased
* Fixed the current unit tests

- [ ] Add unit tests for disabled users

MorrisJobke committed:
optimize SVGs

MorrisJobke commented:
> SVGs need proper viewBox

Why?

rullzer commented:
@MorrisJobke because else firefox et all might desicde to do weird stuff again ;)

MorrisJobke commented:
> @MorrisJobke because else firefox et all might desicde to do weird stuff again ;)

But there is a viewbox ;)

rullzer commented:
now there is ;)

MorrisJobke committed:
change group id to _disabledUsers

MorrisJobke committed:
adjust tests to have at least one disabled user

MorrisJobke committed:
add unit tests for disable method

MorrisJobke committed:
add unit tests for enable method

MorrisJobke commented:
@rullzer Unit tests are added.


This is ready for review - cc @LukasReschke @nickvergessen @rullzer

nickvergessen commented on settings/Controller/UsersController.php:
```diff
@@ -245,15 +242,18 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 		if ($this->isAdmin) {
-
-			if($gid !== '') {
+			if($gid !== '' && $gid !== '_disabledUsers') {
```
Should we add a public constant for this?

MorrisJobke replied:
I would say that this is something for a different PR, because then the "_everyone" group should also be handled.

rullzer commented:
LGTM then!

MorrisJobke commented:
@LukasReschke please review :)

MorrisJobke commented:
@nickvergessen Mind to review this? ;)

MorrisJobke commented:
@LukasReschke I would really love to see this reviewed :)

ChristophWurst commented:
Is it intended that users disappear from the ``Everyone`` list once they get disabled?

MorrisJobke commented:
> Is it intended that users disappear from the Everyone list once they get disabled?

I think everyone should contain all users. Because currently disabled is only a kind of group.

MariusBluem commented:
...Maybe we should introduce something like **Active Users** and **All users** 😁

MorrisJobke commented:
> ...Maybe we should introduce something like Active Users and All users

Could we get this in and then refine afterwards?

nickvergessen commented on settings/Controller/UsersController.php:
```diff
@@ -490,6 +494,135 @@ public function destroy($id) {
+			);
+		}
+
+		if(!$this->isAdmin && !$this->groupManager->getSubAdmin()->isUserAccessible($this->userSession->getUser(), $user)) {
```
`$user` can be null, causing an issue here.

nickvergessen commented on settings/Controller/UsersController.php:
```diff
@@ -490,6 +494,135 @@ public function destroy($id) {
+						'message' => (string)$this->l10n->t('Unable to disable user.')
+					)
+				)
+			);
```
missing status code here. Should not be OK

MorrisJobke replied:
That's how the current JS side works. Changing this would need to redo a lot of the JS code, because it only handles the 200 response code results.

nickvergessen replied:
well other errors do have a non 200 status code...

MorrisJobke replied:
Really? O.o

nickvergessen commented on settings/Controller/UsersController.php:
```diff
@@ -490,6 +494,135 @@ public function destroy($id) {
+			);
+		}
+
+		if(!$this->isAdmin && !$this->groupManager->getSubAdmin()->isUserAccessible($this->userSession->getUser(), $user)) {
```
same here, non existing user breaks

MorrisJobke replied:
I will look into this. You commented twice on the same line. Is this correct?

nickvergessen replied:
No, two lines, once enable and once disable

nickvergessen commented on settings/Controller/UsersController.php:
```diff
@@ -490,6 +494,135 @@ public function destroy($id) {
+					'data' => array(
+						'message' => (string)$this->l10n->t('Unable to enable user.')
+					)
+				)
```
missing status code

MorrisJobke replied:
Same as above. This is how the current JS code works. Not nice, but that's it - we can't redo everything always ;)

MorrisJobke committed:
check $user object before using it

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
@nickvergessen I adjusted the if statements to be before the usage of `$user`. Review time :)

MorrisJobke commented:
I rebased to fix the merge conflicts (only in the tests)

MorrisJobke committed:
fix unit tests

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
@blizzz I fixed the unit tests.

blizzz commented:
If the full name col greyed out intentionally? If find it irritating.

MorrisJobke commented:
> If the full name col greyed out intentionally? If find it irritating.

What? This isn't correct. Which browser?

jancborchardt commented on settings/Controller/UsersController.php:
```diff
@@ -491,6 +494,135 @@ public function destroy($id) {
+				array(
+					'status' => 'error',
+					'data' => array(
+						'message' => (string)$this->l10n->t('Unable to disable user.')
```
This wording is confusing. Better: »Error while disabling user.«

jancborchardt commented on settings/Controller/UsersController.php:
```diff
@@ -491,6 +494,135 @@ public function destroy($id) {
+				array(
+					'status' => 'error',
+					'data' => array(
+						'message' => (string)$this->l10n->t('Unable to disable user.')
```
Same here: »Error while disabling user.«

jancborchardt commented on settings/Controller/UsersController.php:
```diff
@@ -491,6 +494,135 @@ public function destroy($id) {
+				array(
+					'status' => 'error',
+					'data' => array(
+						'message' => (string)$this->l10n->t('Unable to enable user.')
```
»Error while enabling user.«

jancborchardt commented on settings/Controller/UsersController.php:
```diff
@@ -491,6 +494,135 @@ public function destroy($id) {
+				array(
+					'status' => 'error',
+					'data' => array(
+						'message' => (string)$this->l10n->t('Unable to enable user.')
```
»Error while enabling user.«

jancborchardt commented on settings/css/settings.css:
```diff
@@ -242,15 +242,16 @@ table.grid {
+td.password>img,td.displayName>img, td.quota>img { visibility:hidden; }
 td.password, td.quota, td.displayName { width:12em; cursor:pointer; }
-td.password>span, td.quota>span, rd.displayName>span { margin-right: 1.2em; color: #C7C7C7; }
+td.password>span, td.quota>span, td.displayName>span { margin-right: 1.2em; color: #C7C7C7; }
```
The greying out of the display name comes from this change I guess. ;)

jancborchardt commented on settings/js/users/users.js:
```diff
@@ -804,6 +800,54 @@ $(document).ready(function () {
+						GroupList.setUserCount(GroupList.getGroupLI('_disabledUsers'), count+1);
+					}
+				} else {
+					OC.dialogs.alert(result.data.message, t('settings', 'Unable to change status of {user}', {user: uid}));
```
»Error while changing status of …«

jancborchardt commented on tests/Settings/Controller/UsersControllerTest.php:
```diff
@@ -2076,4 +2145,432 @@ public function testSetDisplayNameFails() {
+			[
+				'status' => 'error',
+				'data' => [
+					'message' => 'Unable to disable user.',
```
»Error while disabling user.« – Don’t Repeat Yourself? ;D

jancborchardt commented on tests/Settings/Controller/UsersControllerTest.php:
```diff
@@ -2076,4 +2145,432 @@ public function testSetDisplayNameFails() {
+			[
+				'status' => 'error',
+				'data' => [
+					'message' => 'Unable to disable user.',
```
And again here

jancborchardt commented on tests/Settings/Controller/UsersControllerTest.php:
```diff
@@ -2076,4 +2145,432 @@ public function testSetDisplayNameFails() {
+			[
+				'status' => 'error',
+				'data' => [
+					'message' => 'Unable to disable user.',
```
Just search and replace for »Unable to« ;D

MorrisJobke committed:
fix greyed out display name

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke committed:
Improve wording of error messages

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke committed:
use proper return codes and handle failure cases

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
@ChristophWurst I fixed the layout issue. Thanks @jancborchardt 

@jancborchardt I fixed the wording \o/

@nickvergessen I added the proper return codes and fixed the JS handling of those cases. ;)

blizzz commented:
When I as group admin disable a user,

* [ ] a counter appears to the Disabled group entry, despite having LDAP  enabled
* [ ] the disabled user belonging the the group does not appear within Disbled (== it is empty).

blizzz commented:
When I as group admin disable a user,

* [ ] a counter appears to the Disabled group entry, despite having LDAP  enabled
* [ ] the disabled user belonging the the group does not appear within Disbled (== it is empty).

(cannot add a review??)

rullzer commented:
Conflicts => needs rebase
@MorrisJobke @Faldon can you take a look at @blizz https://github.com/nextcloud/server/pull/1234#issuecomment-252881240

blizzz commented:
>  @MorrisJobke @Faldon can you take a look at @blizz #1234 (comment)

I shall writer @rulzer with a single l now. :imp: ;)

MorrisJobke commented:
Rebased.

MorrisJobke commented:
@blizzz Could you elaborate on your comments. I don't understand them fully.

enoch85 commented:
Wow, yet another very nice improvement. Thanks guys!

MorrisJobke commented:
Sadly will not make it into 11 -> moved to 12

blizzz commented:
rebase needed

nickvergessen commented:
This branch has conflicts that must be resolved

nickvergessen commented:
You forgot to push @blizzz

MorrisJobke commented:
> rebase needed

Done. Mostly because of some major restructuring in the tests class.

Let's see

blizzz commented:
and tests failing

MorrisJobke committed:
fix unit tests

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
> and tests failing

Fixed

blizzz commented:
Almost. Also a signoff check fails.

LukasReschke commented:
Rebased on current master. Since the commits with the missing sign-off were before we added the sign-off requirement was added I'd be tempted to ignore that.

Will review if tests do pass.

MorrisJobke commented:
> Will review if tests do pass.

They do - only sign-off is the problem here :)

blizzz commented:
@MorrisJobke 

> When I as group admin disable a user,
>    a counter appears to the Disabled group entry, despite having LDAP enabled

Seems fixed

>  the disabled user belonging the the group does not appear within Disbled (== it is empty).

1. Before disabling the users:

![spectacle h12133](https://cloud.githubusercontent.com/assets/2184312/21317770/d958b20c-c606-11e6-8161-3a91289e812c.png)

2. After disabling the users, the "Disabled" user count stays at 3 (with LDAP group counts are disabled, because they are expensive, so this magic group is the only one with a count. Not consistent).

(A reload later the group count here is at 5)

3. Going to the Disabled group it does not show any users:

![spectacle z12133](https://cloud.githubusercontent.com/assets/2184312/21317899/61dca8f4-c607-11e6-9ec6-60545ab73380.png)

blizzz reviewed (changes requested):
cf. https://github.com/nextcloud/server/pull/1234#issuecomment-267991153

Luticus commented:
It seems like you guys are already working on a fix that's near complete, but I've been simply using the extrasecurity app  with a "disabled" group to add the feature.  I went looking again because nextcloud 11 seems to have broken extrasecurity and that app doesn't get updates all too often.  I found this thread and thought I'd suggest just creating a "disable user" group similar to the global "everyone" and then simply have authentication logic that checks for membership?  Seems like it may be easier and there's no need for an interface change.  

Just thought I'd put the idea out there.

georgehrke commented:
needs a rebase

MorrisJobke commented:
@ChristophWurst Could you help me out with this one? It's basically fixing some JS stuff? That would be nice :)

oparoz commented on settings/Controller/UsersController.php:
```diff
@@ -503,6 +506,139 @@ public function destroy($id) {
+
+		if($userId === $id) {
+			return new DataResponse(
+				array(
```
Should be replaced by "[" everywhere.

ChristophWurst replied:
done

oparoz replied:
Thanks :)

ChristophWurst committed:
Use short array syntax

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented:
rebased, resolved conflicts and addressed @oparoz´es comment

blizzz commented:
"Disabled" counter is still not being updated, and shown when it probably should not. But users under "disabled" are now shown correctly.

nickvergessen commented:
@MorrisJobke whats the status here?

MorrisJobke commented:
> @MorrisJobke whats the status here?

No time yet - let me quickly look into this.

MorrisJobke commented:
> @MorrisJobke whats the status here?

Sorry ... is not that easy and there are too many other tasks on my list - either we get it in and fix the counter stuff afterwards or we delay this 😕

MorrisJobke commented:
I rebased to fix the conflicts.

> "Disabled" counter is still not being updated, and shown when it probably should not. But users under "disabled" are now shown correctly.

@ChristophWurst Could you look into this and try to fix this? That would be super nice. This PR is laying already around since ages. And I currently have no time :/

ChristophWurst commented:
> @ChristophWurst Could you look into this and try to fix this? That would be super nice. This PR is laying already around since ages. And I currently have no time :/

Sure, but unlikely that I'll find time for it this week.

blizzz commented:
>> @ChristophWurst Could you look into this and try to fix this? That would be super nice. This PR is laying already around since ages. And I currently have no time :/

> Sure, but unlikely that I'll find time for it this week.

Then let's get this in and fix it during beta cycle. It's not a fundamental issue after all.

blizzz commented on settings/Controller/UsersController.php:
```diff
@@ -441,55 +444,187 @@ public function destroy($id) {
+		$userId = $this->userSession->getUser()->getUID();
+		$user = $this->userManager->get($id);
+
+		if ($userId === $id) {
```
The checks could be consolidated instead of being duplicated in `enable()`` and ``disable()`` methods.. Would not see it as a blocker this stage,  but it's better to get this done during beta.

blizzz commented on settings/users.php:
```diff
@@ -92,6 +95,17 @@
 	$subAdmins = false;
 }
+$disabledUsers = 0;
+foreach (OC_User::getUsers() as $uid) {
```
This is bad. We iterate over hundreds of thousands of users for a value that we can get with a simple SQL query (``SELECT COUNT(userid) from oc_preferences where  appid = 'core' AND configkey='enabled' and configvalue = 'false'``). Currently this leads to a bad performance on opening that Users page.

blizzz reviewed (changes requested):
Found a performance regression

blizzz commented:
I take care of ^

blizzz committed:
UserManager can now count disabled users

Users page takes advantage of that

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
consolidate setEnabled method

and fix a unit test

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
Done with the things I pointed at with my last review. While testing my changes I also found the counter to work :) But the menu appears only on the top-right of the table (did not touch JS though).

rullzer commented:
@blizzz so ready to get reviewed and merged?

blizzz commented:
> But the menu appears only on the top-right of the table

@rullzer this is left.

MariusBluem commented:
Just 2 small comments, which also can be fixed later - as they are design 😉 

- Icons for Everyone, Disabled and Enabled (Checkbox Icon, X-Icon, for Everyone I would recommend a list icon)

- Introduce a space between Everyone, Disabled, Enabled and the Groups (like apps management between All apps, activated, ... and the categories from app store) 

@jancborchardt

blizzz committed:
align popovermenu next to … buttonthingy


Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
>   But the menu appears only on the top-right of the table

Merde! Thought all the time it's a JS issue with some wrong placement or whatever… instead, stupid CSS. 

@rullzer ready to get reviewed

@MariusBluem ack to both, but not by me ( ❤️ backend ❤️ ) and perhaps in a new PR 😼

MorrisJobke commented:
Rebased to resolve conflicts in test file

MorrisJobke commented:
Tested and works 👍

MorrisJobke commented:
Nearly 9 months after the rebase - sorry that it took so long, @Faldon
